piff/kernel.py

```diff
@@ -0,0 +1,301 @@
+# Copyright (c) 2016 by Mike Jarvis and the other collaborators on GitHub at
+# https://github.com/rmjarvis/Piff  All rights reserved.
+#
+# Piff is free software: Redistribution and use in source and binary forms
+# with or without modification, are permitted provided that the following
+# conditions are met:
+#
+# 1. Redistributions of source code must retain the above copyright notice, this
+#    list of conditions and the disclaimer given in the accompanying LICENSE
+#    file.
+# 2. Redistributions in binary form must reproduce the above copyright notice,
+#    this list of conditions and the disclaimer given in the documentation
+#    and/or other materials provided with the distribution.
+
+"""
+.. module:: kernel
+"""
+
+import numpy as np
+from scipy.spatial.distance import cdist
+from sklearn.gaussian_process.kernels import StationaryKernelMixin, NormalizedKernelMixin, Kernel
+from sklearn.gaussian_process.kernels import Hyperparameter
+from sklearn.gaussian_process.kernels import _check_length_scale
+
+
+class ExplicitKernel(StationaryKernelMixin, NormalizedKernelMixin, Kernel):
+    """ A kernel that wraps an arbitrary python function.
+
+    :param  fn:  String that can be combined with 'lambda du,dv:' to eval into a lambda expression.
+                 For example, fn="np.exp(-0.5*np.sqrt(du**2+dv**2)/0.1**2)" would make a Gaussian
+                 kernel with scale length of 0.1.
+    """
+    def __init__(self, fn):
+        self.fn = fn
+
+        self.kwargs = {
+            'fn': fn
+        }
+
+        self._fn = self._eval_fn(fn)
+        self.theta = []
+
+    @staticmethod
+    def _eval_fn(fn):
+        # Some potentially useful imports for the eval string.
+        import math
+        import numpy
+        import numpy as np
+        return eval("lambda du,dv:" + fn)
```

Comment:
this could be upgraded to arbitrary numbers of dimensions / arbitrary input variable names if you required the lambda bits to be a part of the fn string as well. This would also let you put in predefined functions which may not have du, dv as a part of them

Reply:
This part was originally wrote by Josh I guess. I just moved it from gp_interp.py to kernel.py. I fond that more convenient to write a separate script where everybody can implement a new kernel in a sklearn style.